app/controllers/utm_links_controller.rb

```diff
@@ -1,14 +1,162 @@
+    end
+
+    save_utm_link(
+      success_message: "Link updated!",
+      error_redirect_path: edit_utm_links_dashboard_path(@utm_link)
+    )
+  end
+
+  def destroy
+    authorize @utm_link
+
+    @utm_link.mark_deleted!
+    redirect_to utm_links_dashboard_index_path(index_route_params.except(:ids).compact), notice: "Link deleted!"
+  end
+
+  def unique_permalink
+    authorize UtmLink, :new?
+
+    render json: { permalink: UtmLink.generate_permalink }
   end
 
   private
     def set_title
       @title = "UTM Links"
     end
+
+    def set_utm_link
+      @utm_link = current_seller.utm_links.find_by_external_id(params[:id])
+      e404 unless @utm_link
+    end
+
+    def index_params
+      {
+        query: params[:query],
+        page: params[:page],
+        sort: extract_sort_params
+      }
+    end
+
+    def index_route_params
+      {
+        query: params[:query],
+        page: params[:page],
+        key: params[:key],
+        direction: params[:direction],
+        ids: params[:ids]
+      }
+    end
+
+    def extract_sort_params
+      key = params[:key]
+      direction = params[:direction]
+
+      return nil unless %w[link date source medium campaign clicks sales_count revenue_cents conversion_rate].include?(key)
+
+      { key: key, direction: direction == "desc" ? "desc" : "asc" }
+    end
+
+    def paginated_utm_links_props
+      PaginatedUtmLinksPresenter.new(
+        seller: current_seller,
+        query: index_params[:query],
+        page: index_params[:page],
+        sort: index_params[:sort]
+      ).props
+    end
+
+    def new_page_props
+      UtmLinkPresenter.new(seller: current_seller).new_page_react_props(copy_from: params[:copy_from])
+    end
+
+    def edit_page_props
+      UtmLinkPresenter.new(seller: current_seller, utm_link: @utm_link).edit_page_react_props
+    end
+
+    def fetch_utm_links_stats(ids)
+      return {} if ids.blank?
+
+      utm_link_ids = current_seller.utm_links.by_external_ids(ids).pluck(:id)
+      UtmLinksStatsPresenter.new(seller: current_seller, utm_link_ids:).props
+    end
+
+    def permitted_params
+      params.require(:utm_link).permit(
+        :title,
+        :target_resource_type,
+        :target_resource_id,
+        :permalink,
+        :utm_source,
+        :utm_medium,
+        :utm_campaign,
+        :utm_term,
+        :utm_content
+      ).merge(
+        ip_address: request.remote_ip,
+        browser_guid: cookies[:_gumroad_guid]
+      )
+    end
+
+    def save_utm_link(success_message:, error_redirect_path:)
+      SaveUtmLinkService.new(
+        seller: current_seller,
+        params: permitted_params,
+        utm_link: @utm_link
+      ).perform
+
+      redirect_to utm_links_dashboard_index_path, notice: success_message, status: :see_other
+    rescue ActiveRecord::RecordInvalid => e
+      error = e.record.errors.first
+      redirect_to error_redirect_path, inertia: { errors: { "utm_link.#{error.attribute}" => error.message } }
```

Comment:
Would this work? What would need to change otherwise?
```suggestion
      error = e.record.errors.first
      redirect_to error_redirect_path, inertia: { errors: e.record.errors }
```

Reply:
No alert flash message?